src/kernels/MIOpenUtilKernels4.cl

```diff
@@ -512,18 +514,18 @@ __kernel void transpose_packed_MN2NM(const global data_t* in,
                                      global data_t* out,
                                      const int N,
                                      const int M,
-                                     const int in_off,
-                                     const int out_off)
+                                     const arg_size_t in_off,
+                                     const arg_size_t out_off)
 {
     uint i = get_global_id(0);
 
     if(i < M * N)
     {
-        uint m_i = iDiv(i, N);
-        uint n_i = iMod(i, m_i, N);
+        const uint m_i = iDiv(i, N);
+        const uint n_i = iMod(i, m_i, N);
 
-        uint in_offset  = m_i * N + n_i + in_off;
-        uint out_offset = n_i * M + m_i + out_off;
+        const size_t in_offset  = m_i * N + n_i + in_off;
+        const size_t out_offset = n_i * M + m_i + out_off;
```

Comment:
Is it what do you want?
`size_t = ulong( uint( (uint_m_i * uint( int_M )) ) + uint_n_i ) + ulong_in_off;`
I mean ulong conversion happens only during the last addition, all the intermediate calculations happened in `uint32` domain.
Also `size_t` in opencl depends on CL_DEVICE_ADDRESS_BITS and can be 32 or 64, and `long` is always a 64 bit values.
Btw, there is `cl_ulong` type which is also 64bit value and it can be a good replacement of `arg_size_t`. But in common case that sort of interoperability can hardly be maintained.

Reply:
@CAHEK7 Thanks for comments.
> Is it what do you want?
> size_t = ulong( uint( (uint_m_i * uint( int_M )) ) + uint_n_i ) + ulong_in_off;

Exactly. Computations in 32-bit domain save a bit of performance. Everywhere we assume that up to 3 dimensions can me multiplied in 32-bit domain without overflow.
https://github.com/ROCm/MIOpen/blob/c5a2384dc0f29682ed51aeccf9b981dbdf7e058f/src/solver/pooling/forwardNaive.cpp#L125-L131

> Also size_t in opencl depends on CL_DEVICE_ADDRESS_BITS and can be 32 or 64

Yes. It is 64 in all devices we support.

> Btw, there is cl_ulong type which is also 64bit value and it can be a good replacement of arg_size_t. But in common case > that sort of interoperability can hardly be maintained.

This won't help if CL_DEVICE_ADDRESS_BITS is 32 because in this case the device is physically unable to address buffers larger than 4GiB.